python-package/lightgbm/sklearn.py

```diff
@@ -662,17 +668,57 @@ def __init__(
         self._n_classes: int = -1
         self.set_params(**kwargs)
 
+    # scikit-learn 1.6 introduced an __sklearn__tags() method intended to replace _more_tags().
+    # _more_tags() can be removed whenever lightgbm's minimum supported scikit-learn version
+    # is >=1.6.
+    # ref: https://github.com/microsoft/LightGBM/pull/6651
     def _more_tags(self) -> Dict[str, Any]:
         return {
             "allow_nan": True,
             "X_types": ["2darray", "sparse", "1dlabels"],
             "_xfail_checks": {
                 "check_no_attributes_set_in_init": "scikit-learn incorrectly asserts that private attributes "
                 "cannot be set in __init__: "
-                "(see https://github.com/microsoft/LightGBM/issues/2628)"
+                "(see https://github.com/microsoft/LightGBM/issues/2628)",
+                "check_n_features_in_after_fitting": (
+                    "validate_data() was first added in scikit-learn 1.6 and lightgbm"
+                    "supports much older versions than that"
+                ),
```

Comment:
On the `1.6.dev` nightlies, `scikit-learn` is raising this error:

> E               AssertionError: `LGBMRegressor.predict()` does not check for consistency between input number
E               of features with LGBMRegressor.fit(), via the `n_features_in_` attribute.
E               You might want to use `sklearn.utils.validation.validate_data` instead
E               of `check_array` in `LGBMRegressor.fit()` and LGBMRegressor.predict()`. This can be done
E               like the following:
E               from sklearn.utils.validation import validate_data

We should ignore this check here in LightGBM... `validate_data()` will be added for the first time in `scikit-learn` 1.6:

* https://github.com/scikit-learn/scikit-learn/blob/74a33757c8a8df84d227f28bbc9ec7ae2fb51dea/sklearn/utils/validation.py#L2790
* https://github.com/scikit-learn/scikit-learn/pull/29696

We have other mechanisms further down in LightGBM to check shape mismatches between training data and the data provided at scoring time. I'd rather rely on those than take on the complexity of try-catching a call to this new-in-v1.6 `validate_data()` function.

Reply:
Understand you don't want to use `validate_data` here, but you can still conform to the API with your own tools.

You probably also want to make sure you store `n_feature_in_` as well, to better imitate sklearn's behavior.

I would personally go down the `fixes.py` path though.

Reply:
> *Understand you don't want to use validate_data here, but you can still conform to the API with your own tools.*

How could we avoid the `check_n_features_in_after_fitting` check failing without calling `validate_data()`? Could you point to a doc I could reference?

> *You probably also want to make sure you store `n_feature_in_` as well, to better imitate sklearn's behavior.*

We do.

https://github.com/microsoft/LightGBM/blob/41ba9e8f00c89d72e5cb71c964722ce1ed4d8445/python-package/lightgbm/sklearn.py#L1063-L1068